File: leetcode/105-从前序与中序遍历序列构造二叉树/105-从前序与中序遍历序列构造二叉树.go
package main

/**
* Definition for a binary tree node.

 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 105-从前序与中序遍历序列构造二叉树
func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 || len(inorder) == 0 {
		return nil
	}
	root := &TreeNode{Val: preorder[0]}
	index := 0
	for i, val := range inorder {
		if val == preorder[0] {
			index = i
			break
		}
	}
	preMidd:= len(inorder[:index]) + 1
	root.Left = buildTree(preorder[1:preMidd], inorder[:index+1])
	root.Right = buildTree(preorder[preMidd:], inorder[index+1:])
	return root
}

File: leetcode/105-从前序与中序遍历序列构造二叉树/105-从前序与中序遍历序列构造二叉树.js
/**
 * Definition for a binary tree node.
 *  */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */

//  105-从前序与中序遍历序列构造二叉树
var buildTree = function (preorder, inorder) {
    if (!inorder.length) {
        return null;
    }
    let root = new TreeNode(preorder[0]);
    let index = inorder.indexOf(preorder[0]) >= 0 ? inorder.indexOf(preorder[0]) : 0;

    root.left = buildTree(preorder.slice(1, index + 1), inorder.slice(0, index));
    root.right = buildTree(preorder.slice(index + 1), inorder.slice(index + 1));
    return root;
};